Extensions "What's New" page updates (#1714)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/whatsnew/index.md

Check this page often to learn about changes to the Chrome extensions platform, its documentation,
and related policy or other changes.

### Chrome 96: Dynamic content scripts {: #m96-dynamic-content-scripts }

The [`chrome.scripting`](/docs/extensions/reference/scripting/) API now supports
[registering](/docs/extensions/reference/scripting/#method-registerContentScripts),
[updating](/docs/extensions/reference/scripting/#method-updateContentScripts),
[unregistering](/docs/extensions/reference/scripting/#method-unregisterContentScripts), and [getting
a list](/docs/extensions/reference/scripting/#method-getRegisteredContentScripts) of content scripts
at runtime. Previously, content scripts could only be statically declared in an extension's
manifest.json or programmatically injected at runtime with
[`chrome.scripting.executeScript()`](/docs/extensions/reference/scripting/#method-executeScript).

> ### Chrome 96: Massively expanded promise support {: #m96-promise-support }

Reviewer: I meant to catch this before it was merged. In numbers what do 'massively' and 'significantly more' actually mean? Did you double the number of promises? Higher? 

The suggestion was described as 'fun'. The problem is that one bit of fun in an otherwise stoic information dump has a danger of being read a hyperbole. We need to avoid that.

cc: @rdcronin

Reviewer 2: More than doubled, I think.

How about:

Expanded promise support to 27 more APIs

Reviewer: Sounds good to me.
```suggestion
### Chrome 96: Expanded promise support to 27 more APIs {: #m96-promise-support }
```

### Chrome 94: Declarative net request static ruleset changes

The [`chrome.declarativeNetRequest`](/docs/extensions/reference/declarativeNetRequest/) now supports
specifying up to 50 static rulesets
([MAX_NUMBER_OF_STATIC_RULESETS](/docs/extensions/reference/declarativeNetRequest/#property-MAX_NUMBER_OF_STATIC_RULESETS))
and enabling up to 10 rulesets
([MAX_NUMBER_OF_ENABLED_STATIC_RULESETS](/docs/extensions/reference/declarativeNetRequest/#property-MAX_NUMBER_OF_ENABLED_STATIC_RULESETS))
at a time.

### Chrome 93: Cross origin isolation support

Both [Manifest V2](/docs/extensions/mv2/cross-origin-isolation/) and [Manifest
V3](/docs/extensions/mv3/cross-origin-isolation/) extensions can now opt into [cross origin
isolation](https://web.dev/cross-origin-isolation-guide/). This feature limits which cross-origin
resources can load an extension's pages and enables the use of low level web platform features like
[`SharedArrayBuffer`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer). Opt in will be required starting in Chrome 95.

### 2021.08.04: Two step verification enforcement {: #two-step-verification-enforcement }

> The [policy update blog post](/blog/policy-update-2sv/) published on 2021-06.29 has been updated to

Reviewer 2: ```suggestion
The [policy update blog post](/blog/policy-update-2sv/) published on 2021.06.29 has been updated to
```

Check this page often to learn about changes to the Chrome extensions platform, its documentation,
and related policy or other changes.

### Chrome 96: Dynamic content scripts {: #m96-dynamic-content-scripts }

The [`chrome.scripting`](/docs/extensions/reference/scripting/) API now supports
[registering](/docs/extensions/reference/scripting/#method-registerContentScripts),
[updating](/docs/extensions/reference/scripting/#method-updateContentScripts),

> [unregistering](/docs/extensions/reference/scripting/#method-unregisterContentScripts), and [getting
> a list of content scripts](/docs/extensions/reference/scripting/#method-getRegisteredContentScripts)

Reviewer 2: ```suggestion
[unregistering](/docs/extensions/reference/scripting/#method-unregisterContentScripts), and [getting
a list of](/docs/extensions/reference/scripting/#method-getRegisteredContentScripts) content scripts
```

Author: Ack.

Reviewer: Please keep the link on 'content scripts'. It's more descriptive than 'list'. Alternatively, you could link the whole phrase.

Author: I'm not sure if this is what @rdcronin had in mind, but as I think about these links a bit more I think I see a clear rationale for keeping "content scripts" outside of the link. This sentence provides a list of operations that are all related to content scripts, and each one links to a relevant method. By scoping the links to the operation and having "content scripts" outside of the link text, it's more obvious to the reader that each of the items in the list operates on the term that follows the list.

At the moment I have a slight to moderate preference for this structure:

> The [`chrome.scripting`] API now supports [registering], [updating], [unregistering], and [getting
a list of] content scripts at runtime.

Reviewer 2: > This sentence provides a list of operations that are all related to content scripts, and each one links to a relevant method. By scoping the links to the operation and having "content scripts" outside of the link text, it's more obvious to the reader that each of the items in the list operates on the term that follows the list.

Yep, exactly this : )

at runtime. Previously, content scripts could only be statically declared in an extension's
manifest.json or programmatically injected at runtime with

> [`chrome.scripting.executeScript`](/docs/extensions/reference/scripting/#method-executeScript).

Author: Ack.

> ### Chrome 96: Expanded promise support {: #m96-promise-support }

Reviewer 2: We increased support here by some non-trivial factor.  Let's add some kind of qualifier - "Significantly expanded" for less fun, "Wildly expanded" (or similar) for more fun.

Author: Happy to! I think in my original draft the intro for this section I said something like "This release contains more promise updates than any previous release." I think I'll change this header and tweak the intro paragraph to lean more into this.

A number of general and Chrome OS-specific extensions APIs have gained promise support in this

> release. Expand the following sections for more details.

Author: Ack.

{% Details %}
{% DetailsSummary %}
Extensions APIs
{% endDetailsSummary %}

> A number of APIs now have promise support in Manifest V3.

Reviewer: ```suggestion
A number of APIs now support promises in Manifest V3.
```

Author: Ack.

> A number of APIs now have promise support in Manifest V3.

Reviewer: Are these classes, methods, or properties? A heuristic I use is to only use 'API' for references to multiple item types. (Language should always be specific.) 

If any of these are methods they should have '()' after them.

Reviewer 2: These are all "namespaces", but we generally refer to them as "APIs" - e.g. the tabs API, the history API, etc.  This is consistent with our language elsewhere.

Reviewer: 'Namespaces' is something I never would have guessed. A writing best practice has always been to prefer the more specific term. In an electronic world, a more specific term results in fewer search results. It's an aid to finding information.

Reviewer 2: "APIs" is the canonical term for these.

- [`chrome.browsingData`](/docs/extensions/reference/browsingData)
- [`chrome.commands`](/docs/extensions/reference/commands)
- [`chrome.contentSettings`](/docs/extensions/reference/contentSettings/)
- [`chrome.debugger`](/docs/extensions/reference/debugger)
- [`chrome.downloads`](/docs/extensions/reference/downloads)
- [`chrome.enterprise.hardwarePlatform`](/docs/extensions/reference/enterprise_hardwarePlatform)
- [`chrome.fontSettings`](/docs/extensions/reference/fontSettings)
- [`chrome.history`](/docs/extensions/reference/history)
- [`chrome.instanceID`](/docs/extensions/reference/instanceID)
- [`chrome.permissions`](/docs/extensions/reference/permissions)
- [`chrome.processes`](/docs/extensions/reference/processes)
- [`chrome.search`](/docs/extensions/reference/search)
- [`chrome.sessions`](/docs/extensions/reference/sessions)
- [`chrome.signedInDevices`](/docs/extensions/reference/signedInDevices)
- [`chrome.topSites`](/docs/extensions/reference/topSites)

Also, APIs that use the [`ChromeSetting`](/docs/extensions/reference/types/#type-ChromeSetting)

> prototype now also support promises. The following APIs are affected by this change.

Reviewer: Same question here and throughout.

- [`chrome.privacy`](/docs/extensions/reference/privacy/)
- [`chrome.accessibilityFeatures`](/docs/extensions/reference/accessibilityFeatures/)
- [`chrome.proxy`](/docs/extensions/reference/proxy/)

{% endDetails %}

{% Details %}
{% DetailsSummary %}
Chrome OS APIs
{% endDetailsSummary %}

- [`chrome.certificateProvider`](/docs/extensions/reference/certificateProvider)
- [`chrome.documentScan`](/docs/extensions/reference/documentScan)
- [`chrome.enterprise.deviceAttributes`](/docs/extensions/reference/enterprise_deviceAttributes)
- [`chrome.enterprise.networkingAttributes`](/docs/extensions/reference/enterprise_networkingAttributes)
- [`chrome.fileBrowserHandler`](/docs/extensions/reference/fileBrowserHandler)
- [`chrome.fileSystemProvider`](/docs/extensions/reference/fileSystemProvider)
- [`chrome.loginState`](/docs/extensions/reference/loginState)
- [`chrome.printingMetrics`](/docs/extensions/reference/printingMetrics)
- [`chrome.wallpaper`](/docs/extensions/reference/wallpaper)

{% endDetails %}

### Chrome 96: `declarativeNetRequestWithHostAccess` permission

> The new `declarativeNetRequestWithHostAccess` permission allows existing Manifest V2 extensions that

Author: Thanks for the clarification and suggestion. I wanted to include this because it wasn't clear to me from the dNR docs why or when I would use this. I quite like this suggestion. I'd also to integrate this language (or something like it) into the dNR API docs in a separate PR.

> leverage `webRequest`, `webRequestBlocking`, and site-specific host permission to migrate to the

Reviewer: ```suggestion
use `webRequest`, `webRequestBlocking`, and site-specific host permission to migrate to the
```

Reviewer: The tech community likes to use 'leverage' as a synonym for 'use'. Unfortunately non-English speakers (and folks new to tech) won't find this definition in a dictionary and translations may or may not handle it correctly.

Author: Ack.

[`chrome.declarativeNetRequest`](/docs/extensions/reference/declarativeNetRequest/) API without

> requiring the end user to approve a new permissions grant.

Reviewer: ```suggestion
requiring the user to approve a new permissions grant.
```

Author: Ack.

> requiring the end user to approve a new permissions grant.

Reviewer 2: ```suggestion
requiring the user to approve new permissions.
```

Author: Ack.

> ### 2021-10-27: Review violation troubleshooting updates {: #2021-10-27-reivew-troubleshooting }

Reviewer 2: We're inconsistent in our date format here - some use 2021-10-27 and some use 2021.10.27.

I personally like the "look" of 2021.10.27 better, but that's purely aesthetic.

Author: Oops, thanks for catching this. I was moving from the period separator to the dash because I'm not used to seeing dates in this format. That's personal preference, though.

At the moment I think it's better to avoid unnecessary churn and go back to periods.

The [Troubleshooting Chrome Web Store violations](/docs/webstore/troubleshooting/) page has been
updated to provide developers with more detailed guidance for common reasons for rejection.

### Chrome 95: Inject scripts directly into pages {: #m95-page-script-injection }

> The [`chrome.scripting`](/docs/extensions/reference/scripting) API's

Reviewer: >The `chrome.scripting` API's

Please be specific. What is an API called in the Extensions universe? Class? Object? Interface?

Reviewer: Please look at this throughout. This isn't just pickiness. You're missing an opportunity to provide information readers who are new to Extensions.

[`executeScript()`](/docs/extensions/reference/scripting/#method-executeScript) method can now

> inject scripts directly into a page's main world. Previously, extensions could only inject directly

Reviewer: Regarding 'main world', is there a more technical term for this? 'Main process?'

Reviewer: You misunderstood the question. 'Main world' doesn't sound like any programmer speak I've ever heard. I wanted to make sure this is actually what it's called.

Reviewer 2: "Main world" and "isolated world" are the canonical terms.

> into the extension's isolated world. For more information on isolated worlds, see the documentation

Reviewer: >isolated world

Same question.

on [content scripts](/docs/extensions/mv3/content_scripts/#isolated_world).

### Chrome 95: Promise support for Storage API {: #m95-storage-promise-support }

Methods on the Manifest V3 version of the [`chrome.storage`](/docs/extensions/reference/storage/)
API now return promises.

### Chrome 95: Cross origin isolation opt-in now required {: #m95-coi-required }

> Extensions and apps must now opt into [cross origin

Reviewer 2: ```suggestion
Extensions and apps should now opt into [cross origin
```

We had to roll back the enforcement on this.  But we should still document / encourage developers to move.

Reviewer 2: Actually, since we cover this below, maybe we just remove this...

Author: Roger, removing for now.

More generally, I'd like to get better at noting when a feature becomes available in Canary, but that would have required more digging into individual issues and commit history than I was up for (or should do at all) last night.

> isolation](https://web.dev/cross-origin-isolation-guide/) in order to use features like

Reviewer: ```suggestion
isolation](https://web.dev/cross-origin-isolation-guide/) to use features like
```

Author: Dropping this section per [@rdcronin's comment](https://github.com/GoogleChrome/developer.chrome.com/pull/1714/files#r746789544).

> SharedArrayBuffer. Documentation on this feature is available for both [Manifest

Reviewer: ```suggestion
`SharedArrayBuffer`. Documentation on this feature is available for both [Manifest
```

Author: Dropping this section per [@rdcronin's comment](https://github.com/GoogleChrome/developer.chrome.com/pull/1714/files#r746789544).

V2](/docs/extensions/mv2/cross-origin-isolation/) and [Manifest

> V3](/docs/extensions/mv3/cross-origin-isolation/) is available.

Reviewer: ```suggestion
V3](/docs/extensions/mv3/cross-origin-isolation/).
```

Author: Dropping this section per [@rdcronin's comment](https://github.com/GoogleChrome/developer.chrome.com/pull/1714/files#r746789544).

### 2021-09-23: Manifest V2 support timeline {: #manifest-v2-support-timeline }

The Manifest V2->V3 transition timeline was [announced in this blog post](/blog/mv2-transition/) and
a more detailed [timeline page](/docs/extensions/mv3/mv2-sunset) was published.

### Chrome 94: Declarative net request static ruleset changes

The [`chrome.declarativeNetRequest`](/docs/extensions/reference/declarativeNetRequest/) API static

> ruleset implementation has been adjusted in order to give developers more flexibility. Previously

Reviewer: ```suggestion
ruleset has been adjusted to give developers more flexibility. Previously
```

Reviewer: Removal of 'implementation'. Again, this may seem minor. Distinguishing the implementation from the design is only relevant to audiences who deal with both. To a user of Extensions, it's not an implementation. It's just the way it works.

Author: Dropping this language in favor of [@rdcronin's suggestion](https://github.com/GoogleChrome/developer.chrome.com/pull/1714/files#r746792080).

developers could only specify and enable up to 10 rulesets. Developers can now specify up to 50
rulsets
([MAX_NUMBER_OF_STATIC_RULESETS](/docs/extensions/reference/declarativeNetRequest/#property-MAX_NUMBER_OF_STATIC_RULESETS))
and enable up to 10 rulsets
([MAX_NUMBER_OF_ENABLED_STATIC_RULESETS](/docs/extensions/reference/declarativeNetRequest/#property-MAX_NUMBER_OF_ENABLED_STATIC_RULESETS))

> at a time.

Author: Thanks for both the cleanup and typo fix. I realize now that while I wanted to explain the change, the original intent of this page was to provide developers with a running log of changes. Our plan was that for cases where additional information is necessary, this page would link out to other resources.

### Chrome 93: Cross origin isolation support

Both [Manifest V2](/docs/extensions/mv2/cross-origin-isolation/) and [Manifest
V3](/docs/extensions/mv3/cross-origin-isolation/) extensions can now opt into [cross origin
isolation](https://web.dev/cross-origin-isolation-guide/). This feature limits which cross-origin
resources can load an extension's pages and enables the use of low level web platform features like

> [SharedArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer). Opt in will be required starting in Chrome 95.

Reviewer: ```suggestion
[`SharedArrayBuffer`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer). Opt in will be required starting in Chrome 95.
```

Author: Ack.

### 2021-06-29: Developer program policies updated {: #developer-program-policies-updated }

The Chrome Web Store [Developer Program Policies](/docs/webstore/program_policies) have been updated
with clarifications to the deceptive installation tactics, spam, and repetitive content policies.
This update also includes a new two step verification requirement to publish on the Chrome Web
Store. [Read the blog post](/blog/policy-update-2sv/) for more information.

### 2021.06.23: "Extension actions in Manifest V3" blog post {: #new-blog-post-extension-actions-in-manifest-v3 }

Chrome extensions had `chrome.browserAction` and `chrome.pageActions` APIs for years, but Manifest
V3 replaced both with a generic [`chrome.actions`](/docs/extensions/reference/action/) API. This
post explores the history of these APIs and what has changed in Manifest V3. [Read the
post](/blog/mv3-actions).

### 2021.06.08: "Introducing chrome.scripting" blog post {: #new-blog-post-introducing-chromescripting }

The [`chrome.scripting`](/docs/extensions/reference/scripting/) API is a new Manifest V3 API focused
on, well, scripting. In this post we dig into the motivations for this change and take a closer look

> at some of the new capabilities it introduces. [Read the post](/blog/crx-scripting-api).

Reviewer: >at some of the new capabilities it introduces

This could be tighter:

- at it's new capabilities.

or 

- at capabilities it introduces.

Author: Ack. Using "at it's new capabilities".

### Chrome 91: Module service worker support {: #es-modules-for-service-workers }

> Chrome now supports JavaScript modules in service workers. In your manifest, specify a module in
> your manifest:

Reviewer: ```suggestion
Chrome now supports JavaScript modules in service workers. To specify a module in
your manifest:
```

The new [`chrome.scripting.removeCSS()`](/docs/extensions/reference/scripting/#method-removeCSS)
method allows extensions to remove CSS that was previously inserted via
[`chrome.scripting.insertCSS()`](/docs/extensions/reference/scripting/#method-insertCSS). It
replaces [`chrome.tabs.removeCSS()`](/docs/extensions/reference/tabs/#method-removeCSS).

### Chrome 90: `chrome.scripting.executeScript()` results include frameId {: # chromescriptingexecutescript-results-include-frameid }

Results returned from
[`chrome.scripting.executeScript()`](/docs/extensions/reference/scripting/#method-executeScript)
now include the [frameId](/docs/extensions/reference/webNavigation/#a-note-about-frame-ids).
The `frameId` property indicates the frame that the result is from, letting extensions easily
associate results with the individual frames when injecting in multiple frames.

### Chrome 89: New API for managing tab groups {: #new-api-for-tab-groups-mv3-only }

The new [`chrome.tabGroups`](/docs/extensions/reference/tabGroups/) API and new capabilities in
[`chrome.tabs`](/docs/extensions/reference/tabs/) let extensions read and manipulate tab groups.
Requires Manifest V3.

> ### Chrome 89: Customizable permissions for Web Accessible Resources {: #customizable-permissions-for-mv3-web-accessible-resources }

Reviewer: ```suggestion
### Chrome 89: Customizable permissions for web accessible resources {: #customizable-permissions-for-mv3-web-accessible-resources }
```

Author: Ack.
